torch_geometric/nn/conv/heat_pyg.py

```diff
@@ -0,0 +1,187 @@
+    return [True if x in wanted_types else False for x in node_or_edge_type]
+
+def flatten_a_list_of_list(list_of_list):
+    flat_list = []
+    for l in list_of_list:
+        flat_list += l
+    return flat_list
+
+class HEATlayer(nn.Module):
+    ''' 
+        1.  type-specific transformation for nodes of different types: ('n1', 'n2').
+            transform nodes from different vector space to the same vector space.
+        2.  edges are assumed to have different types but contains the same attributes.            
+    '''
+    def __init__(self,  node_type_set = ('n1', 'n2', 'n3'), edge_type_set = ('e1', 'e2', 'e3', 'e4'),
+                        in_channels_node=4, in_channels_edge_attr=2, in_channels_edge_type=2, 
+                        edge_attr_emb_size=5, edge_type_emb_size=5, node_emb_size=5, 
+                        out_channels=6, heads=3, concat=True):
+
+        super(HEATlayer, self).__init__()
+        ## Parameters
+        self.node_type_set = node_type_set
+        self.edge_type_set = edge_type_set
+        self.in_channels_node = in_channels_node # 32
+        self.in_channels_edge_attr = in_channels_edge_attr # 2
+        self.in_channels_edge_type = in_channels_edge_type # 2
+        self.out_channels = out_channels
+        self.heads = heads
+        self.concat = concat
+        self.edge_attr_emb_size = edge_attr_emb_size
+        self.edge_type_emb_size = edge_type_emb_size
+        self.node_emb_size = node_emb_size
+
+        #### Layers ####
+        ## Embeddings 
+        self.set_node_emb()
+        self.set_edge_emb()
+
+        ## Transform the concatenated edge_nbrs feature to out_channels to update the next node feature
+        self.node_update_emb = nn.Linear(self.edge_attr_emb_size + self.node_emb_size, self.out_channels, bias=False) 
+
+        ## Attention
+        self.attention_nn = nn.Linear(self.edge_attr_emb_size + self.edge_type_emb_size + 2*self.node_emb_size, 1*self.heads, bias=False)
+        self.leaky_relu = nn.LeakyReLU(0.2)
+        self.soft_max = nn.Softmax(dim=1)
+    
+    def set_node_emb(self):
+        ''' assume that different nodes have the same dimension, but different vector space. '''
+        for node_type in self.node_type_set:
+            exec('self.{}_node_feat_emb = nn.Linear(self.in_channels_node, self.node_emb_size, bias=False)'.format(node_type))
+    
+    def set_edge_emb(self):
+        ''' assume that different edges have different types but contains the same kind of attributes. '''
+        # for edge_type in self.edge_type_set:
+        #     exec('self.{}_edge_type_emb = nn.Linear(self.in_channels_edge_type, self.edge_type_emb_size, bias=False) '.format(edge_type))
+        self.edge_type_emb = nn.Linear(self.in_channels_edge_type, self.edge_type_emb_size, bias=False) 
+        self.edge_attr_emb = nn.Linear(self.in_channels_edge_attr, self.edge_attr_emb_size, bias=False) 
+
+    def embed_nodes(self, node_features, node_type_list):
```

Comment:
Can we make this a general module outside of `HEATConv`, *.e.g*, `HeteroLinear(torch.nn.Module)`? This should be helpful for all kinds of heterogeneous graph learning.

Reply:
In this `embed_nodes` function, we are embedding node features according to their types. so in this function, we need the predefined embedding functions (which you suggest implementing with ModuleList) that should be predefined for each task to initialize HEATConv. Do you mean we pass the ModuleList as an argument to the HeteroLinear (embed_nodes) function?